packages/contracts-bedrock/src/dispute/OutputBisectionGame.sol

```diff
@@ -0,0 +1,551 @@
+        uint256 _splitDepth,
+        Duration _gameDuration,
+        IBigStepper _vm
+    ) {
+        if (_splitDepth >= _maxGameDepth) revert InvalidSplitDepth();
+
+        GAME_TYPE = _gameType;
+        ABSOLUTE_PRESTATE = _absolutePrestate;
+        GENESIS_BLOCK_NUMBER = _genesisBlockNumber;
+        MAX_GAME_DEPTH = _maxGameDepth;
+        SPLIT_DEPTH = _splitDepth;
+        GAME_DURATION = _gameDuration;
+        VM = _vm;
+    }
+
+    ////////////////////////////////////////////////////////////////
+    //                  `IOutputBisectionGame` impl               //
+    ////////////////////////////////////////////////////////////////
+
+    /// @inheritdoc IOutputBisectionGame
+    function step(uint256 _claimIndex, bool _isAttack, bytes calldata _stateData, bytes calldata _proof) external {
+        // INVARIANT: Steps cannot be made unless the game is currently in progress.
+        if (status != GameStatus.IN_PROGRESS) revert GameNotInProgress();
+
+        // Get the parent. If it does not exist, the call will revert with OOB.
+        ClaimData storage parent = claimData[_claimIndex];
+
+        // Pull the parent position out of storage.
+        Position parentPos = parent.position;
+        // Determine the position of the step.
+        Position stepPos = parentPos.move(_isAttack);
+
+        // INVARIANT: A step cannot be made unless the move position is 1 below the `MAX_GAME_DEPTH`
+        if (stepPos.depth() != MAX_GAME_DEPTH + 1) revert InvalidParent();
+
+        // Determine the expected pre & post states of the step.
+        Claim preStateClaim;
+        ClaimData storage postState;
+        if (_isAttack) {
+            // If the step position's index at depth is 0, the prestate is the absolute
+            // prestate.
+            // If the step is an attack at a trace index > 0, the prestate exists elsewhere in
+            // the game state.
+            preStateClaim = stepPos.indexAtDepth() == 0
+                ? ABSOLUTE_PRESTATE
+                : findTraceAncestor(Position.wrap(Position.unwrap(parentPos) - 1), parent.parentIndex).claim;
+            // For all attacks, the poststate is the parent claim.
+            postState = parent;
+        } else {
+            // If the step is a defense, the poststate exists elsewhere in the game state,
+            // and the parent claim is the expected pre-state.
+            preStateClaim = parent.claim;
+            postState = findTraceAncestor(Position.wrap(Position.unwrap(parentPos) + 1), parent.parentIndex);
+        }
+
+        // INVARIANT: The prestate is always invalid if the passed `_stateData` is not the
+        //            preimage of the prestate claim hash.
+        //            We ignore the highest order byte of the digest because it is used to
+        //            indicate the VM Status and is added after the digest is computed.
+        if (keccak256(_stateData) << 8 != Claim.unwrap(preStateClaim) << 8) revert InvalidPrestate();
```

Comment:
This seems like it would be a useful library function

Reply:
Will address in follow up